build_runner/lib/src/entrypoint/run_script.dart

```diff
@@ -0,0 +1,145 @@
+
+  @override
+  String get invocation =>
+      '${super.invocation.replaceFirst('[arguments]', '[build-arguments]')} '
+      '<executable> [-- [script-arguments]]';
+
+  @override
+  SharedOptions readOptions() {
+    // The default option parser will throw if we pass additional arguments,
+    // because it expects positional arguments to be build directories.
+    //
+    // Instead, pass an empty list, and we'll handle positional arguments ourselves.
+    return new SharedOptions.fromParsedArgs(
+        argResults, [], packageGraph.root.name, this);
+  }
+
+  @override
+  FutureOr<int> run() async {
+    // Ensure that the user passed the name of a file to run.
+    if (argResults.rest.isEmpty) {
+      stderr
+        ..writeln('Must specify an executable to run.')
+        ..writeln()
+        ..writeln(usage);
+      return ExitCode.usage.code;
+    }
+
+    var scriptName = argResults.rest[0];
+    var passedArgs = argResults.rest.skip(1).toList();
+    var options = readOptions();
+
+    // Create a temporary directory in which to execute the script.
+    var tempPath = Directory.systemTemp
+        .createTempSync('build_runner_run_script')
+        .absolute
+        .uri
+        .toFilePath();
+
+    // Create two ReceivePorts, so that we can quit when the isolate is done.
+    //
+    // Define these before starting the isolate, so that we can close
+    // them if there is a spawn exception.
+    ReceivePort onExit, onError;
+
+    try {
+      var outputMap = options.outputMap ?? {};
+      outputMap.addAll({tempPath: null});
+
+      var result = await build(
+        builderApplications,
+        deleteFilesByDefault: options.deleteFilesByDefault,
+        enableLowResourcesMode: options.enableLowResourcesMode,
+        configKey: options.configKey,
+        assumeTty: options.assumeTty,
+        outputMap: outputMap,
+        packageGraph: packageGraph,
+        verbose: options.verbose,
+        builderConfigOverrides: options.builderConfigOverrides,
+        isReleaseBuild: options.isReleaseBuild,
+        trackPerformance: options.trackPerformance,
+        skipBuildScriptCheck: options.skipBuildScriptCheck,
+        buildDirs: options.buildDirs,
+        logPerformanceDir: options.logPerformanceDir,
+      );
+
+      if (result.status == BuildStatus.failure) {
+        stdout.writeln('Skipping script run due to build failure');
+        return result.failureType.exitCode;
+      }
+
+      // Find the path of the script to run.
+      var scriptPath = p.setExtension(p.join(tempPath, scriptName), '.dart');
+      var packageConfigPath = p.join(tempPath, '.packages');
+
+      // Use a completer to determine the exit code.
+      var completer = new Completer<int>();
+
+      Isolate isolate;
+      onExit = new ReceivePort();
+      onError = new ReceivePort();
+
+      // On an error, kill the isolate, and rethrow the error.
+      onError.listen((e) {
+        isolate?.kill();
```

Comment:
Are we sure we want this to exit on unhandled errors? I don't know if its clear either way what should be done.

Reply:
In my head, it makes sense to kill the isolate after it crashes, mainly since the running script would have stopped executing, and we printed the error to the screen.

I can't think of a reason where you'd want to keep the isolate alive (plus it's probably already dead by this point), but maybe a flag (like `--no-kill-isolate-on-error`) might work out?

Reply:
Should I mark this as resolved? IMO it makes sense to leave it this way.

Reply:
Yes, I think so. I'll send a commit in right now.

Reply:
Just LMK if there's anything else I should add or fix up!

Reply:
We'll take another look soon - sorry for the delay.

Reply:
Since we're using `errorsAreFatal` now I think we can remove this line